epub33/rs/index.html

```diff
@@ -1170,24 +1146,124 @@ <h2>Open Container Format Processing</h2>
 				<h3>OCF Abstract Container</h3>
 
 				<section id="sec-container-iri">
-					<h4>Relative URLs for Referencing Other Components</h4>
+					<h4>URL of the Root Directory</h4>
 
-					<p>For <a href="https://url.spec.whatwg.org/#relative-url-with-fragment-string"
-							>relative-URL-with-fragment strings</a> [[URL]], Reading Systems MUST determine the <a
-							href="https://url.spec.whatwg.org/#concept-base-url">base URL</a> [[URL]] according to the
-						relevant language specifications for the given file formats. For example, CSS defines how
-						relative URL references work in the context of CSS style sheets and property declarations
-						[[CSSSnapshot]].</p>
+					<p>Reading Systems MUST assign a URL [[URL]] to the <a>Root Directory</a> of the <a>OCF Abstract Container</a>. This URL is called the <a data-cite="epub-33#dfn-container-root-url">container root URL</a>. The URL itself is implementation specific, but:</p>
+
+					<ul>
+						<li>The result of <a data-cite="url#concept-url-parser">parsing </a> [[URL]] "<code>/</code>" with the <a>container root URL</a> as <a data-cite="url#concept-base-url">base</a> MUST be the <a>container root URL</a>.</li>
+						<li>The result of <a data-cite="url#concept-url-parser">parsing </a> [[URL]] "<code>..</code>" with the <a>container root URL</a> as <a data-cite="url#concept-base-url">base</a> MUST be the <a>container root URL</a>.</li>
+						<li>The <a data-cite="url#origin">origin</a> of the <a>container root URL</a> MUST be unique for each EPUB Publication instance.</li>
+						<li>the <a data-cite="url#origin">origins</a> of the <a data-cite="epub-33#dfn-container-url">container URLs</a> of all the files in the <a>OCF Abstract Container</a> MUST be identical.</li>	
```

Comment:
I am a little uneasy with this requirement. Are we really certain there are no situations where a RS might want/need to use different origins?

Reply:
I do not really see any such case... I would propose leave this as is now and, once we have merged, we would raise an issue asking this very question to the community. We can always come back to this requirement later.

Reply:
Off the top of my head, a RS that supports spine level scripting might put all files identified as scripted in one origin, but place all non-scripted spine items in unique origins. For that matter, unique origins likely only matter for scripted content so it doesn't seem critical to include it here. But I am fine accepting these edits and opening an issue once it is in.

Reply:
> unique origins likely only matter for scripted content so it doesn't seem critical to include it here

Right, I wanted to make a similar comment: origin-related requirements aren't critical to the original issue of knowing how to parse URLs in EPUB.

It is critical however to how RS are supposed to implement the Root URL. Depending on these origin-related criteria, an implementation may be conforming or not.

These criteria may create conformance issues to existing RS, while not solving a perfectly clear user issue. They might be on the "theoretical purity" side of the priority of constituencies.

@iherman instead of adding them and discuss later, wouldn't it be safer to drop them and discuss later? unless we hear from RS implementors of course.

Reply:
> Right, I wanted to make a similar comment: origin-related requirements aren't critical to the original issue of knowing how to parse URLs in EPUB.

I do hope we can find some way to retain the idea of origin, so that we have a foundation for thinking about security issues.

Reply:
> I do hope we can find some way to retain the idea of origin, so that we have a foundation for thinking about security issues.

Oh yes, I totally agree it's a foundational issue. It's just that it can be separated from the current PR, and requires review/consensus from RS people.

Reply:
Are the two criteria of the same weight? I certainly feel that the first one (ie, the unicity of the origin for the root URL) is important to keep and, actually, it is the result of an earlier WG resolution. So I would not want to remove that and make it the subject of a separate issue, @rdeltour. (Or we reinstate the relevant entry that I removed in the section on scripting...)

I am less sure about the 4th entry: I would still prefer to keep the PR as is in this respect. I am not sure I understand the logic of an RS doing what @bduga describe, ie, why an RS would want to do that and how would it affect the behavior of EPUB instances. That is why I would prefer to stick with my original proposal: stick with the PR as is, and discuss it later via a separate issue. By keeping it we have put a stake on the ground that _something_ must be said about this...

Reply:
> Are the two criteria of the same weight?

I don't think they do.

> I certainly feel that the first one (ie, the unicity of the origin for the root URL) is important to keep and, actually, it is the result of an earlier WG resolution. So I would not want to remove that and make it the subject of a separate issue, @rdeltour. (Or we reinstate the relevant entry that I removed in the section on scripting…)

OK, I agree.

> I am less sure about the 4th entry: I would still prefer to keep the PR as is in this respect. I am not sure I understand the logic of an RS doing what @bduga describe, ie, why an RS would want to do that and how would it affect the behavior of EPUB instances. 

Yeah the fate of that 4th conformance statement is less obvious.
One issue I can see is that it puts a lot of constraint on the Root URL implementation chosen by the RS. Notably, all custom-scheme URLs (`epub:` and the likes, `httpsr2:`, `ibooks-epub:`) would be disallowed: the origin of each custom-scheme URL is a new opaque origin (see the "otherwise" clause of the [URL’s origin definition](https://url.spec.whatwg.org/#origin)). That could make a lot of RS non-conforming, for a requirement that is not necessarily backed by a strong use case.

You'd think I would actually defend my original proposal 😅, but since this is a spec PR and not a mere issue comment, I'm more in favor of dropping this 4th statement for now.

With the 3d statement, the origin stake is on the ground. We know that something _can_ be said along the lines of the 4th statement, but also that it _may_ be too restrictive. We're not exactly paving any cowpath, nor answering a clearly stated use case. So I'd vote for moving that discussion to an issue rather than preemptively making the spec stricter than it should.

Reply:
@rdeltour, I find this argument

> all custom-scheme URLs (epub: and the likes, httpsr2:, ibooks-epub:) would be disallowed: the origin of each custom-scheme URL is a new opaque origin (see the "otherwise" clause of the URL’s origin definition).

compelling indeed. So I will remove the last restriction of the text in the next commit.